### PR TITLE
gh-64376: Convert the time module to the Argument Clinic

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Co-Authored-By: Nikolaus Rath <nikolaus@rath.org>

https://bugs.python.org/issue20177

* Issue: gh-64376

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
As mentioned on the BPO issue, it doesn't seem practical to use the Argument Clinic instead of the `gettmarg()` utility function.

The `time_clock_gettime()` function was not converted -- there's an `ifdef` involving the `PyArg_ParseTuple()` call on AIX.

File: Lib/test/test_gdb.py [hunk 1]
```diff
@@ -849,7 +849,7 @@ def test_pycfunction(self):
         # Various optimizations multiply the code paths by which these are
         # called, so test a variety of calling conventions.
         for py_name, py_args, c_name, expected_frame_number in (
-            ('gmtime', '', 'time_gmtime', 1),  # METH_VARARGS
+            ('gmtime', '', 'time_gmtime', 2),  # METH_VARARGS
```
Comment (jdemeyer) [inline Lib/test/test_gdb.py]: This change is not correct. The whole point of this test is that we're testing a `METH_VARARGS` function. See #14330 for a fix.

**Review by jdemeyer — CHANGES_REQUESTED** [review]
Needs #14330

File: Lib/test/test_gdb.py (hunk 1 above)
Comment (ZackerySpytz) [reply Lib/test/test_gdb.py]: I'm sorry. I did not realize that the Argument Clinic would use `METH_FASTCALL` for `gmtime()`.

**Comment by jdemeyer** [comment]
Needs #14795 instead.

**Comment by encukou** [comment]
There are now (in current master) dedicated built-in functions to use for testing, which aren't very likely to gain Argument Clinic signatures. I suggest using `_testcapi.meth_varargs`, which returns  (its module, tuple of arguments):

```python
        import _testcapi
        meth = types.MethodType(_testcapi.meth_varargs, 32)
        self.assertEqual(meth(), (_testcapi, (32,)))
```

**Comment by jdemeyer** [comment]
> There are now (in current master) dedicated built-in functions to use for testing, which aren't very likely to gain Argument Clinic signatures. I suggest using `_testcapi.meth_varargs`, which returns (its module, tuple of arguments):

This PR shouldn't change `test_gdb.py` at all. The only reason that it was changed in this PR was to fix a test failure because `test_gdb.py` was using `gmtime()`.

**Comment by encukou** [comment]
It no longer changes `test_gdb`. My comment was about `Lib/unittest/test/testmock/testhelpers.py`.
Sorry for the confusion.

File: Modules/timemodule.c [hunk 2]
```diff
@@ -418,37 +436,41 @@ tmtotuple(struct tm *p
...
+        return 0;
     }
-    *pwhen = whent;
     return 1;
 }
 
+/*[clinic input]
+time.gmtime
+
+    seconds: object = NULL
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c:454]: Maybe use a converter?
```suggestion
    seconds as when: object(converter="object_to_time_t", type="time_t") = None
```
Or introduce a named converter?

Note also that you can use `None` as a default value for better signature.

File: Modules/timemodule.c [hunk 3]
```diff
@@ -842,45 +866,70 @@ time_strftime(PyObject *self, PyObject *args)
...
-See the library reference manual for formatting codes. When the time tuple\n\
-is not present, current time as returned by localtime() is used.\n\
-\n" STRFTIME_FORMAT_CODES);
 #endif /* HAVE_STRFTIME */
 
+/*[clinic input]
+time.strptime
+
+    string: object
+    format: object = NULL
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c:875]:
```suggestion
    format: object(c_default="NULL") = "%a %b %d %H:%M:%S %Y"
```

File: Modules/timemodule.c [hunk 4]
```diff
@@ -842,45 +866,70 @@ time_strftime(PyObject *self, PyObject *args)
...
+the C library strftime function.
+[clinic start generated code]*/
+
 static PyObject *
-time_strptime(PyObject *self, PyObject *args)
+time_strptime_impl(PyObject *module, PyObject *string, PyObject *format)
+/*[clinic end generated code: output=00d2a25afc00983a input=474fd69ff9188bcb]*/
 {
-    PyObject *module, *func, *result;
+    PyObject *mod, *func, *result;
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c:908]: Why rename `module`?

File: Modules/timemodule.c [hunk 5]
```diff
@@ -1282,13 +1365,17 @@ time_process_time(PyObject *self, PyObject *unused)
...
-PyDoc_STRVAR(process_time_doc,
-"process_time() -> float\n\
-\n\
-Process time for profiling: sum of the kernel and user-space CPU time.");
+/*[clinic input]
+time.process_time_ns
+
+Process time for profiling as nanoseconds:
+
+sum of the kernel and user-space CPU time.
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c:1373]: It is a single sentence. Either write it in one line or make separate sentences.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

**Comment by taleinat** [comment]
Ping, @ZackerySpytz?

**Comment by erlend-aasland** [comment]
No response since 2020. In the three years since 2020, there has been a lot of changes to Modules/timemodule.c[^1]. I suggest to close this PR; if someone wants to pick this up, it will probably be easier to start from scratch, rather than rebasing this PR.

[^1]: 569 insertions(+), 231 deletions(-)

**Pull request closed without merging**